transitions/utils.js

```diff
@@ -41,44 +38,27 @@ module.exports = {
     module.exports.addRejectionMessage(document, reportConfig, 'registration_not_found');
   },
   isIdUnique: function(db, id, callback){
-    utils.getRegistrations({
-        db: db,
-        id: id
-    }, function(err, registrations) {
-        if (err) {
-            callback(err);
-        } else if (registrations.length) {
-            callback(null, false);
-        } else {
-            callback(null, true);
-        }
+    db.medic.view('medic', 'registered_patients', {
+      key: id
+    }, (err, registrations) => {
+      if (err) {
+          callback(err);
+      } else if (registrations.length) {
+          callback(null, false);
+      } else {
+          callback(null, true);
+      }
     });
   },
   addUniqueId: function(db, doc, callback) {
-    var potentialIds = _.map(Array(IDS_TO_GENERATE), _.partial(ids.generate, currentIdLength));
-
-    utils.getRegistrations({
-        db: db,
-        ids: potentialIds
-    }, function(err, registrations) {
-        if (err) {
-            return callback(err);
-        }
-
-        var uniqueIds = _.reject(potentialIds, function(id) {
-          return _.find(registrations, function(registration) {
-            return registration.key === id;
-          });
-        });
-
-        if (!uniqueIds.length) { // id collision, retry
-            logger.warn('Could not create a unique id of length ' + currentIdLength + ', increasing by one');
-            currentIdLength += 1;
-            module.exports.addUniqueId(db, doc, callback);
-        } else {
-            doc.patient_id = uniqueIds[0];
-            callback();
-        }
+    console.log('addUniqueId');
+    idGenerator.next().value.then(patientId => {
```

Comment:
I get what you're saying. Seems like a bit of a toss up though honestly. In this implementation, `ids` is a stateless module that provides a "factory" (remember that word?) to generate a stateful iterator. The downside being, whomever wants that iterator has to iterate over it. In your version `ids` would hold that state. It's just where you want the state to sit, and thus where the iteration occurs.

I am personally OK with it being where it is. I think the ugliness that you're worried about is actually just inherent to JS generators (I wish `.next()` just got you the value, but I guess they didn't want logic to be confused around whether or not the iterator was done due to falsy. Perhaps they should have `hasNext()` and `next()` for those who need that distinction). If you want to play around with moving it feel free to give it a crack.